swupd/create_manifests.go

```diff
@@ -382,29 +364,32 @@ func CreateManifests(version, previous, minVersion uint32, format uint, statedir
 	// copy over unchanged manifests
 	addUnchangedManifests(&newMoM, oldMoM, groups)
 
-	// allManifests must include newManifests plus all old ones in the MoM.
-	allManifests, err := aggregateManifests(newManifests, &newMoM, version, c)
-	if err != nil {
-		return nil, err
-	}
-
 	var osIdx *Manifest
-	if osIdx, err = writeIndexManifest(&c, &ui, allManifests); err != nil {
+	if osIdx, err = writeIndexManifest(&c, &ui, oldMoM, newManifests); err != nil {
 		return nil, err
 	}
 
-	// read index manifest from the version directory specified in the manifest
-	// itself as an index manifest may not have been created for this version.
-	osIdxDir := filepath.Join(c.outputDir, fmt.Sprint(osIdx.Header.Version))
-	osIdxPath := filepath.Join(osIdxDir, "Manifest."+osIdx.Name)
-	if err = newMoM.createManifestRecord(osIdxDir, osIdxPath, osIdx.Header.Version, TypeManifest, osIdx.BundleInfo.Header.Status); err != nil {
-		return nil, err
-	}
+	if osIdx != nil {
+		// read index manifest from the version directory specified in the manifest
+		// itself as an index manifest may not have been created for this version.
+		osIdxDir := filepath.Join(c.outputDir, fmt.Sprint(osIdx.Header.Version))
+		osIdxPath := filepath.Join(osIdxDir, "Manifest."+osIdx.Name)
+		if err = newMoM.createManifestRecord(osIdxDir, osIdxPath, osIdx.Header.Version, TypeManifest, osIdx.BundleInfo.Header.Status); err != nil {
+			return nil, err
+		}
 
-	// track here as well so the manifest tar is made
-	// but only if we made it new for this version
-	if osIdx.Header.Version == newMoM.Header.Version {
-		newManifests = append(newManifests, osIdx)
+		// track here as well so the manifest tar is made
+		// but only if we made it new for this version
+		if osIdx.Header.Version == newMoM.Header.Version {
+			newManifests = append(newManifests, osIdx)
+		}
+	} else {
+		// Carry old index if present
+		for _, f := range oldMoM.Files {
+			if f.Name == IndexBundle {
+				newMoM.Files = append(newMoM.Files, f)
```

Comment:
Can there be multiple entries for IndexBundle in oldMoM? If not, we could `break` here.

Reply:
Fixed.

Reply:
Fixed